alicloud/data_source_alicloud_security_groups.go

```diff
@@ -0,0 +1,168 @@
+package alicloud
+
+import (
+	"fmt"
+	"log"
+	"regexp"
+
+	"github.com/denverdino/aliyungo/ecs"
+	"github.com/denverdino/aliyungo/util"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+type SecurityGroup struct {
+	Attribute    ecs.DescribeSecurityGroupAttributeResponse
+	CreationTime util.ISO6801Time
+}
+
+func dataSourceAlicloudSecurityGroups() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAlicloudSecurityGroupsRead,
+
+		Schema: map[string]*schema.Schema{
+			"name_regex": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"output_file": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			// Computed values
+			"groups": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vpc_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"inner_access": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+						"creation_time": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceAlicloudSecurityGroupsRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AliyunClient).ecsconn
+
+	regionId := getRegion(d, meta)
+
+	args := &ecs.DescribeSecurityGroupsArgs{
+		RegionId: regionId,
+	}
+
+	var sg []SecurityGroup
+
+	var nameRegex *regexp.Regexp
+	if v, ok := d.GetOk("name_regex"); ok {
+		if r, err := regexp.Compile(v.(string)); err == nil {
+			nameRegex = r
+		}
+	}
+
+	for {
+		items, paginationResult, err := conn.DescribeSecurityGroups(args)
+		if err != nil {
+			return fmt.Errorf("DescribeSecurityGroups: %#v", err)
+		}
+
+		for _, item := range items {
+			if v, ok := d.GetOk("vpc_id"); ok && v.(string) != item.VpcId {
+				continue
```

Comment:
Line 100 may be needless and ecs.DescribeSecurityGroupsArgs support 'VpcId'.

Reply:
fixed